Add support for google sheets

KobbyMmo opened the pull request:
## Description

Allow frontend to pull and store google sheets data into local storage

Fixes # (N/A)

## Type of change

- [x] New feature (non-breaking change which adds functionality)


## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

KobbyMmo committed:
add support for google sheets

kilemensi commented on public/dashboard/index.html:
```diff
@@ -33,7 +35,7 @@
         <a href="#footer" class="nav-link-5 w-nav-link">PARTNERS</a>
         <div class="select-form w-form">
-          <form id="wf-form-select-form" name="wf-form-select-form"   data-name="select form" class="select"><select onchange="zoomMap(this.value)" id="county-select" name="county-select" data-name="county-select"  class="select-field-3 w-select">
+          <form id="wf-form-select-form" name="wf-form-select-form"   data-name="select form" class="select"><select onchange="locationChanged(this.value)" id="county-select" name="county-select" data-name="county-select"  class="select-field-3 w-select">
```
Use `onXXXChange` or `handleXXXChange` pattern

kilemensi commented on public/js/events.js:
```diff
@@ -1,9 +1,11 @@
   const newUrl = `${baseUrl}${countiesLocation[value].zoom}/${countiesLocation[value].latitude}/${countiesLocation[value].longitude}`;
 
   const countyName = countiesLocation[value].name;
+  const numberOfHouseholds =await gethouseHold(countyName);
```
```suggestion
  const numberOfHouseholds = await getHouseHoldCount(county);
```

kilemensi commented on public/js/sheets.js:
```diff
@@ -0,0 +1,86 @@
+async function loadSheets() {
+  //county-cities
+  Papa.parse(
+    'https://docs.google.com/spreadsheets/d/1jNk90L1FGXt3estVzFII2-eeKQZ85RYiLKCyNe14nGg/export?format=csv&gid=0',
+    {
+      download: true,
+      header: true,
+      complete: function (results) {
+        var countyCityMap = results.data;
+        setLocalItem('countyCityMap', countyCityMap);
+      },
+    }
+  );
+
+  //household-information
+  Papa.parse(
+    'https://docs.google.com/spreadsheets/d/1jNk90L1FGXt3estVzFII2-eeKQZ85RYiLKCyNe14nGg/export?format=csv&gid=679979845',
+    {
+      download: true,
+      header: true,
+      complete: function (results) {
+        var countyHouseholdMap = results.data;
+        setLocalItem('countyHouseholdMap', countyHouseholdMap);
+      },
+    }
+  );
+}
```
Since each sheet must be loaded independently, why don't we load each sheet in its own function wrapped in a [Promise](https://stackoverflow.com/a/56427246) so we can use async. e.g.
```js
   const countyCitiesMap = await loadCountyCitiesMap();
```

KobbyMmo replied:
I might have to move `countyCitiesMap` to the backend, not very useful in the frontend

kilemensi replied:
Sure... if we're not using it in FE, lets leave it to the API.

kilemensi commented on public/js/sheets.js:
```diff
@@ -0,0 +1,86 @@
+  }
+  const citiesInfo = countyCityMap.find((row) =>
+    row.Cities.toLowerCase().include(city.toLowerCase())
+  );
```
```suggestion
  const citiesInfo = countyCityMap.find((row) =>
    row.Cities.toLowerCase().includes(city.toLowerCase())
  );
```
I take it you mean [String.prototype.includes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes)?

kilemensi commented on public/js/sheets.js:
```diff
@@ -0,0 +1,86 @@
+  );
+  if (!householdInfo) {
+    return 0;
+  }
```
```suggestion
  const householdInfo = countyHouseholdMap && countyHouseholdMap.find(
    (row) => row.County.toLowerCase() === county.toLowerCase()
  );
  if (!householdInfo) {
    return 0;
  }
```

... and may be we should `trim` as well since somebody may mistakenly add space(s) to the county name.

kilemensi reviewed (commented):
👍🏾

esirK commented on public/js/sheets.js:
```diff
@@ -0,0 +1,86 @@
+
+async function init() {
+  if (!getLocalItem('countyHouseholdMap') || !getLocalItem('countyCityMap')) {
+    loadSheets();
```
shouldn't we have `await` here? 
`await loadSheets();`

KobbyMmo replied:
we can, but this is called when the DOM is loaded, and nothing has been selected. We don't want to block loading just to wait for sheets. The initial display doesn't really require it.

esirK replied:
Nice. Still learning this stuff

kilemensi commented:
Why is this still unresolved @KobbyMmo ?

KobbyMmo committed:
add papaparse and request

KobbyMmo committed:
pull county information from google sheets

KobbyMmo committed:
refactor; rename variabes, handle inputs

kilemensi commented on public/js/sheets.js:
```diff
@@ -0,0 +1,63 @@
+  ); 
+}
+
+async function gethouseHold(county) {
```
```suggestion
async function getHouseHoldCount(county) {
```

kilemensi commented on public/js/sheets.js:
```diff
@@ -0,0 +1,63 @@
+
+
+
+window.addEventListener('DOMContentLoaded', init);
```
`init is` an _`async`_ function. Does it play nice with event listeners?

KobbyMmo replied:
yes, it does. I can omit the async, but the `loadCountyHouseholdMap` is async anyway. this is to ensure the function is loaded on startup.

kilemensi replied:
As long as you're sure it works, I'm good for it.

kilemensi commented on src/api/index.js:
```diff
@@ -626,6 +628,43 @@ const API = {
 
+async function loadCountyCitiesMap() {
+  return new Promise((resolve, reject) => {
+    const options = {
+      header: true,
+      error(err) {
+        reject(err);
+      },
+    };
+    const dataStream = request.get(
+      'https://docs.google.com/spreadsheets/d/1jNk90L1FGXt3estVzFII2-eeKQZ85RYiLKCyNe14nGg/export?format=csv&gid=0'
+    );
+    const parseStream = Papa.parse(Papa.NODE_STREAM_INPUT, options);
+    dataStream.pipe(parseStream);
```
Can you may be comment (or add a link to specific doc section) why we need to do all these somersaults.

kilemensi reviewed (commented):
👍🏾 @KobbyMmo 

-------------------

Did we figure out how are we going to at least lint the code we write?

KobbyMmo committed:
rename and comment variables

kilemensi reviewed (commented):
Changing household value in the gsheet is not reflected on the dashboard @KobbyMmo

KobbyMmo commented:
> Changing household value in the gsheet is not reflected on the dashboard @KobbyMmo

@kilemensi  value is cached for 3 minutes to avoid always going to the server. [Here](https://github.com/CodeForAfrica/sensors.AFRICA-Dashboard-KE-HAP/pull/8/files#diff-2690e5c4d2cc11c40d1b688319e481ea6ae9a2511675e6c45a5cb6cbc54b1c17R30-R44)

kilemensi commented:
> > Changing household value in the gsheet is not reflected on the dashboard @KobbyMmo
> 
> @kilemensi value is cached for 3 minutes to avoid always going to the server. [Here](https://github.com/CodeForAfrica/sensors.AFRICA-Dashboard-KE-HAP/pull/8/files#diff-2690e5c4d2cc11c40d1b688319e481ea6ae9a2511675e6c45a5cb6cbc54b1c17R30-R44)

I know that @KobbyMmo but I changed it like 10 minutes ago but Nairobi household count still shows as 12 on my end (it should be 22 now).

KobbyMmo commented:
> I know that @KobbyMmo but I changed it like 10 minutes ago but Nairobi household count still shows as 12 on my end (it should be 22 now).

@kilemensi  Looking at the sheet now, there are actually 2 Nairobi in there,  12, and another 22. Removed one, and adding validation in the Sheet

KobbyMmo committed:
fix; call Date.now()
